components/core/src/clp/ffi/Value.hpp

```diff
@@ -0,0 +1,224 @@
+ * needed when the immutable view type is a const reference or some other types.
+ * @tparam Type
+ */
+template <typename Type>
+struct ImmutableViewTypeConverter {
+    using type = Type;
+};
+
+/**
+ * Specializes `std::string`'s immutable view type to `std::string_view`.
+ */
+template <>
+struct ImmutableViewTypeConverter<std::string> {
+    using type = std::string_view;
+};
+
+/**
+ * Specializes `clp::ir::EightByteEncodedTextAst`'s immutable view type its const reference.
+ */
+template <>
+struct ImmutableViewTypeConverter<clp::ir::EightByteEncodedTextAst> {
+    using type = clp::ir::EightByteEncodedTextAst const&;
+};
+
+/**
+ * Specializes `clp::ir::FourByteEncodedTextAst`'s immutable view type its const reference.
+ */
+template <>
+struct ImmutableViewTypeConverter<clp::ir::FourByteEncodedTextAst> {
+    using type = clp::ir::FourByteEncodedTextAst const&;
+};
+
+/**
+ * Template alias to the underlying typename of `ImmutableViewTypeConverter`.
+ * @tparam Type
+ */
+template <typename Type>
+using ImmutableViewType = typename ImmutableViewTypeConverter<Type>::type;
+
+class Value {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(
+                ErrorCode error_code,
+                char const* const filename,
+                int line_number,
+                std::string message
+        )
+                : TraceableException{error_code, filename, line_number},
+                  m_message{std::move(message)} {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return m_message.c_str();
+        }
+
+    private:
+        std::string m_message;
+    };
+
+    // Constructors
+    Value() = default;
+
+    /**
+     * Move constructs from the given rvalue.
+     * @tparam PrimitiveValue The type of the rvalue, which must be a move constructable
+     * non-reference type.
+     * @value
+     */
+    template <MoveConstructableValueType PrimitiveValue>
+    requires(false == std::is_reference_v<PrimitiveValue>)
+    Value(PrimitiveValue&& value) : m_value{std::forward<PrimitiveValue>(value)} {}
+
+    /**
+     * Constructs from the given fundamental-type value.
+     * @tparam PrimitiveValue The type of the given value, which must be a C++ fundamental type.
+     * @value
+     */
+    template <FundamentalValueType PrimitiveValue>
+    Value(PrimitiveValue value) : m_value{value} {}
+
+    // Disable copy constructor and assignment operator
+    Value(Value const&) = delete;
+    auto operator=(Value const&) -> Value& = delete;
+
+    // Default move constructor and assignment operator
+    Value(Value&&) = default;
+    auto operator=(Value&&) -> Value& = default;
+
+    // Destructor
+    ~Value() = default;
+
+    // Methods
+    /**
+     * @tparam PrimitiveValue
+     * @return Whether the underlying value is the given `PrimitiveValue` type.
```

Comment:
```suggestion
     * @tparam T
     * @return Whether the underlying value is the given type.
```

Note the rename of the template parameter.